niworkflows/interfaces/bids.py

```diff
@@ -596,8 +596,16 @@ def _run_interface(self, runtime):
                         # load updated NIfTI
                         nii = nb.load(out_file, mmap=False)
                     data_dtype = np.dtype(data_dtype)
-                    if nii.get_data_dtype() != data_dtype:
+                    orig_dtype = nii.get_data_dtype()
+                    if orig_dtype != data_dtype:
+                        LOGGER.warning(
+                            f"Changing {out_file} dtype from {orig_dtype} to {data_dtype}"
+                        )
+                        # coerce dataobj to new data dtype
+                        new_data = np.asanyarray(nii.dataobj).astype(data_dtype)
```

Comment:
```suggestion
                        new_data = np.asanyarray(nii.dataobj, dtype=data_dtype)
```

Reply:
Is this going to cast BOLD data to int16?

Reply:
i think https://github.com/poldracklab/smriprep/issues/195 is part of the problem, but there is some further step downstream that is upcasting to float. But this is also necessary for any conversion to match `source_file` (currently just `boldref`).

Reply:
@effigies this should only affect data suffixes:
```
        ("mask", "uint8"),
        ("dseg", "int16"),
        ("probseg", "float32"),
        ("boldref", "source")
```
unless the `data_dtype` flag is explicitly checked, which I don't believe it ever is.

Reply:
> i think [poldracklab/smriprep#195](https://github.com/poldracklab/smriprep/issues/195) is part of the problem, but there is some further step downstream that is upcasting to float.

commented there, yes there is a resampling downstream in this case - moving to standard space.